src/commands/tail/mod.rs

```diff
@@ -1,15 +1,51 @@
+use std::net::{SocketAddr, TcpListener};
+
 use crate::settings::global_user::GlobalUser;
 use crate::settings::toml::Target;
 use crate::tail::Tail;
 
-pub fn start(target: &Target, user: &GlobalUser) -> Result<(), failure::Error> {
-    // Note that we use eprintln!() throughout this file; this is because we want any
+const DEFAULT_TUNNEL_PORT: u16 = 8080;
+const DEFAULT_METRICS_PORT: u16 = 8081;
+
+pub fn start(
+    target: &Target,
+    user: &GlobalUser,
+    tunnel_port: Option<u16>,
+    metrics_port: Option<u16>,
+) -> Result<(), failure::Error> {
+    let t_port = find_open_port(tunnel_port, DEFAULT_TUNNEL_PORT)?;
+    let m_port = find_open_port(metrics_port, DEFAULT_METRICS_PORT)?;
+
+    // Note that we use eprintln!() throughout this module; this is because we want any
     // helpful output to not be mixed with actual log JSON output, so we use this macro
     // to print messages to stderr instead of stdout (where log output is printed).
     eprintln!(
-        "Setting up log streaming from Worker script \"{}\" to Wrangler. This may take a few seconds...",
-        target.name
+        "Setting up log streaming from Worker script \"{}\". Using ports {} and {}. This may take a few seconds...",
+        target.name,
+        t_port,
+        m_port,
     );
 
-    Tail::run(target.clone(), user.clone())
+    Tail::run(target.clone(), user.clone(), t_port, m_port)
+}
+
+/// Find open port takes two arguments: an Optional requested port, and a default port.
+fn find_open_port(requested: Option<u16>, default: u16) -> Result<u16, failure::Error> {
+    if let Some(port) = requested {
+        let addr = format!("127.0.0.1:{}", port);
+        match TcpListener::bind(addr) {
+            Ok(socket) => Ok(socket.local_addr()?.port()),
+            Err(_) => failure::bail!("the specified port {} is unavailable", port),
+        }
+    } else {
```

Comment:
if the user did not specify a port, just provision one for  them. try to use a default but don't block on it.

Reply:
beautiful